### ISSUE TITLE
Question related to plugin booting up

### ISSUE THREAD

**Issue by Roky7250** [sysrepo/sysrepo#2127]
Hi, is it possible to somehow affect to the order of plugin booting up?

For example if I have two modules, test_a.so and test_b.so, is it possible to say in which order these two plugins will be booted by sysrepo-plugind?

Regards

**Comment by michalvasko** [comment]
Hi,
no, currently it is not possible. But it should not be difficult to add. You can provide such a patch (for example, a mandatory priority variable in every plugin) or I will add once I finish my current work.

**Comment by maksimvuleta** [comment]
It will be good if have that feature. 
Do have any hints what we need to add to enable the plugin booting up order feature?
I am not so familiar with the sysrepo source code.

**Comment by michalvasko** [comment]
I have provided quite a good hint but if it does not help you at all, you had better leave it up to me.

**Comment by maksimvuleta** [comment]
Ok, thanks. 
We will wait for you to update it.

**Comment by maksimvuleta** [comment]
Any news on this? 
Can we expect the new release with this feature soon?

**Comment by michalvasko** [comment]
I have it in my TODO, but I am not managing to do the standard work so there is absolutely no time for such optional features. So I am sorry, I cannot tell you anything more, it will likely not be soon.

**Comment by maksimvuleta** [comment]
This is a real issue when we have dependencies between yang modules.
For example, LLDP yang module has leaf definition like this:
      _leaf name {
        type if:interface-ref;
        must "derived-from-or-self(deref(.)/../**if:type**, 'ianaift:ethernetCsmacd')" 
           + " or derived-from-or-self(deref(.)/../**if:type**, 'ianaift:ieee8023adLag')" {
          error-message
            "The LLDP is only configured on Link Aggreagation
             and Ethernet Port.";
        }
        description   
          "The port name used to identify the port component
           (contained in the local chassis with the LLDP agent)
           associated with this entry.";
      }_

The point is about **if:type** definition from ietf-interfaces.yang module. 
If LLDP yang is booted up before the ietf-interfaces yang module the sysrepo-plugind boot up will crash because if:type leaf does not exist.
If ietf-interfaces is booted up before LLDP everything will work fine.

**Comment by michalvasko** [comment]
Okay, it should now be supported. Just configure it the desired way in `sysrepo-plugind` YANG configuration.

**Comment by Roky7250** [comment]
Hi, I solved this issue in a simpler way (not saying that it is a better way).
I am using sysrepo with 1.4.51 version (devel branch, commit id = 7784c41fc0e180a329922efd4edd3d01eaa4f1a8).

So I created a patch file for sysrepo-plugind.c and sysrepo.h files and it is attached as [sysrepo_patch.txt](https://github.com/sysrepo/sysrepo/files/5695744/sysrepo_patch.txt).

I added variable **plugin_boot_order** in each module.
I load that variable for each module inside function **load_plugins** to the variable **plugin_boot_order_var**.
Basically, I expanded structure **srpd_plugin_s** with new filed `unsigned int *plugin_boot_order_var`.

After modules are loaded I call the sort function. Function sorts sysrepo-plugins in ascending order using **plugin_boot_order** variable for comparison.

Everything works fine on my side. Boot up order is working as it should.

Is this solution good enough? 

Regards

**Comment by michalvasko** [comment]
That was my original idea as well but `sysrepo-plugind` will in all likelihood have much more functionality (that will need to be configured) in future so the best solution is to keep all configuration together in one place, in the YANG data. Also, this solution may be a but problematic because when implementing a module, you must know about all the other modules that you may depend on/that may depend on you to set the order variable correctly. That is impossible, obviously.

**Comment by maksimvuleta** [comment]
Thank you both!

**Comment by Roky7250** [comment]
Hi, I installed the latest sysrepo from devel.
I have three modules installed. They are called: example-module, second-example-module and third-example-module.

When I do sysrepoctl --list I get:
```
Module Name                | Revision   | Flags | Owner     | Permissions | Submodules | Features
-------------------------------------------------------------------------------------------------
example-module             |            | I     | root:root | 600         |            |
ietf-datastores            | 2018-02-14 | I     | root:root | 666         |            |
ietf-inet-types            | 2013-07-15 | i     |           |             |            |
ietf-netconf               | 2011-06-01 | I     | root:root | 666         |            |
ietf-netconf-notifications | 2012-02-06 | I     | root:root | 666         |            |
ietf-netconf-with-defaults | 2011-06-01 | I     | root:root | 666         |            |
ietf-origin                | 2018-02-14 | I     | root:root | 666         |            |
ietf-yang-library          | 2019-01-04 | I     | root:root | 666         |            |
ietf-yang-metadata         | 2016-08-05 | i     |           |             |            |
ietf-yang-types            | 2013-07-15 | i     |           |             |            |
second-example-module      |            | I     | root:root | 600         |            |
sysrepo-monitoring         | 2020-04-17 | I     | root:root | 600         |            |
sysrepo-plugind            | 2020-12-10 | I     | root:root | 666         |            |
third-example-module       |            | I     | root:root | 600         |            |
yang                       | 2017-02-20 | I     | root:root | 666         |            |
```

I used sysrepocfg to set plugin boot order. My configuration is:
```
<config xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
<sysrepo-plugind xmlns="urn:sysrepo:plugind">
    <plugin-order>
        <plugin>example-module</plugin>
        <plugin>second-example-module</plugin>
        <plugin>third-example-module</plugin>
    </plugin-order>
</sysrepo-plugind>
</config>
```

I executed next command: 
`sudo sysrepocfg --edit=tmp.xml --datastore=startup`

Then I started sysrepo-plugind. 
The order of plugins booting up was not how I configure it.
Am I doing something wrong here?

Regards

**Comment by michalvasko** [comment]
Can you post the output of `ls <plugin-dir>`? So that I can see what plugins you want to execute exactly.

**Comment by Roky7250** [comment]
Hi,
The output of my plugins directory is: `ls -l /usr/lib/plugins`

```
-rw-r--r-- 1 root root 13256 Dec 21 04:52 libfirst-example-module.so
-rw-r--r-- 1 root root 13256 Dec 21 04:52 libsecond-example-module.so
-rw-r--r-- 1 root root 13256 Dec 21 04:52 libthird-example-module.so
```

Regards

**Comment by michalvasko** [comment]
Right, please read the node [description](https://github.com/sysrepo/sysrepo/blob/devel/modules/sysrepo-plugind.yang#L13) and change the values accordingly. One plugin may handle several YANG modules or several plugins can handle subtrees from a single YANG module data.

**Comment by Roky7250** [comment]
Ahh, sorry for my mistake.
It is logical to put the name of the `.so` file related to the YANG file for which the .`so` file is created.
It works fine now.

Once again sorry for my mistake.

Regards
